File: .github/scripts/merge-dotenv.mjs
#!/usr/bin/env node
import fs from 'node:fs';

const keyPattern = /^[A-Za-z_][A-Za-z0-9_]*$/;
const values = new Map();

function parse(file) {
  if (!fs.existsSync(file)) return;
  for (const [index, raw] of fs.readFileSync(file, 'utf8').split(/\r?\n/).entries()) {
    let line = raw.trim();
    if (!line || line.startsWith('#')) continue;
    if (line.startsWith('export ')) line = line.slice(7).trimStart();
    const separator = line.indexOf('=');
    if (separator < 0) throw new Error(`malformed dotenv line ${file}:${index + 1}`);
    const key = line.slice(0, separator).trim();
    let value = line.slice(separator + 1).trim();
    if (!keyPattern.test(key)) throw new Error(`invalid dotenv key ${file}:${index + 1}`);
    if (value.startsWith('"') || value.startsWith("'")) {
      const quote = value[0];
      if (value.length < 2 || !value.endsWith(quote)) {
        throw new Error(`unterminated dotenv value ${file}:${index + 1}`);
      }
      value = value.slice(1, -1);
    }
    values.set(key, value);
  }
}

try {
  for (const file of process.argv.slice(2)) parse(file);
  for (const [key, value] of values) process.stdout.write(`${key}=${value}\n`);
} catch (error) {
  console.error(`dotenv merge: ${error.message}`);
  process.exit(1);
}

File: .github/scripts/merge-dotenv.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

KEY_PATTERN = /\A[A-Za-z_][A-Za-z0-9_]*\z/
values = {}

begin
  ARGV.each do |file|
    next unless File.exist?(file)

    File.readlines(file, chomp: true).each_with_index do |raw, index|
      line = raw.strip
      next if line.empty? || line.start_with?('#')

      line = line.sub(/\Aexport\s+/, '')
      separator = line.index('=')
      raise "malformed dotenv line #{file}:#{index + 1}" unless separator

      key = line[0...separator].strip
      value = line[(separator + 1)..].strip
      raise "invalid dotenv key #{file}:#{index + 1}" unless KEY_PATTERN.match?(key)

      if value.start_with?('"', "'")
        quote = value[0]
        raise "unterminated dotenv value #{file}:#{index + 1}" unless value.length >= 2 && value.end_with?(quote)
        value = value[1...-1]
      end
      values[key] = value
    end
  end

  values.each { |key, value| puts "#{key}=#{value}" }
rescue StandardError => e
  warn "dotenv merge: #{e.message}"
  exit 1
end
